### PR TITLE
music player service

### PR DESCRIPTION
**Pull request by mehdirouhandeh**
This is a service for Exoplayer to run music in the background along with a notification

### CODE REVIEW

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt [hunk 1]
```diff
@@ -0,0 +1,669 @@
...
+        makeAudioManager()
+        resetDataQueue()
+        currentWindow = 0
+        playbackPosition = 0
+        initializePlayer()
+    }
+
+    private fun makeAudioManager() {
+        mAudioManager = getSystemService(Context.AUDIO_SERVICE) as AudioManager
+        if (mAudioManager != null) mAudioManager!!.requestAudioFocus(
```
Comment (IamHaddad) [inline app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt]: request audio focus deprecate shode goya
https://developer.android.com/reference/android/media/AudioManager#requestAudioFocus(android.media.AudioManager.OnAudioFocusChangeListener,%20int,%20int)

Comment (mehdirouhandeh) [reply app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt]: I've fixed the problem, feel free if you have any suggestion to improve the code

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt [hunk 2]
```diff
@@ -0,0 +1,684 @@
...
+        pausedByAudioListener = false
+        player!!.seekTo(currentWindow, playbackPosition)
+        player!!.playWhenReady = true
+        player!!.playbackState
+        showCurrentTrackNotification()
+    }
+
+    private fun initializePlayer() {
+        if (player == null) {
+            player = ExoPlayerFactory.newSimpleInstance(
```
Comment (IamHaddad) [inline app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt]: in kaht hm barresi konid lotfan deprecate shode

Comment (mahdi302) [reply app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt]: ok shod

File: app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListFragment.kt [hunk 3]
```diff
@@ -32,6 +34,8 @@ class MusicsListFragment : Fragment() {
     @Inject
     lateinit var viewModelFactory: ViewModelFactory
 
+    private var musicList = ArrayList<Music>()
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListFragment.kt]: toye fragment?!!!

File: app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListFragment.kt [hunk 4]
```diff
@@ -67,12 +72,20 @@ class MusicsListFragment : Fragment() {
...
+            prepareMusicListService(musics)
             Log.d("tag", it.toString())
             musicListAdapter.submitList(musics)
         }
         musicListViewModel.postMusic().observe(viewLifecycleOwner, musicObserver)
     }
 
+    private fun prepareMusicListService(musics: List<MusicRepoModel>) {
+            musicList.clear()
+        for (musicRepo in musics) {
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListFragment.kt]: jaye in fek nemikonam to fragment bashe

File: app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListFragment.kt [hunk 5]
```diff
@@ -67,12 +72,20 @@ class MusicsListFragment : Fragment() {
         val musicObserver = Observer<List<MusicRepoModel>> {
 
             val musics = it
+            prepareMusicListService(musics)
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListFragment.kt]: kolle list ro pass midim be service?
ke chi beshe? age gharare kolesh pakhsh beshe khob khode service mitone az Repo begire dige nemitone?

File: app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListFragment.kt [hunk 6]
```diff
@@ -128,10 +141,12 @@ class MusicsListFragment : Fragment() {
     }
 
     private fun musicListener(music: Music) {
-        val action = MusicsListFragmentDirections.actionMusicsListFragmentToPlayerFragment(music)
-
-        findNavController().navigate(action)
+//        val action = MusicsListFragmentDirections.actionMusicsListFragmentToPlayerFragment(music)
+//
+//        findNavController().navigate(action)
+        ExoPlayerService.actionStart(context, musicList,music.id)
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListFragment.kt]: in testi inja goazshte shode?!
chera jeloye baz shodane player fragment ro greftim?

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt [hunk 7]
```diff
@@ -0,0 +1,425 @@
...
+    }
+
+    private fun buildMediaSource(uri: Uri?): MediaSource? {
+        val userAgent = "star_player"
+        return ProgressiveMediaSource.Factory(DefaultHttpDataSourceFactory(userAgent))
+            .createMediaSource(uri)
+    }
+
+    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
+        if (intent != null) {
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt]: age null bod chi?

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt [hunk 8]
```diff
@@ -0,0 +1,425 @@
...
+
+
+    private fun play(trackModel: Music?) {
+        val trackModels = ArrayList<Music?>()
+        trackModels.add(trackModel)
+        play(trackModels)
+    }
+
+    private fun play(trackModels: ArrayList<Music?>?) {
+        reInit()
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt]: benazaram behtare moghe sakhte shodan e service yekbar faghat player ro besazim
injori dari hardafe ye player jadid misazi

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt [hunk 9]
```diff
@@ -0,0 +1,425 @@
...
+    }
+
+
+    private fun play(trackModel: Music?) {
+        val trackModels = ArrayList<Music?>()
+        trackModels.add(trackModel)
+        play(trackModels)
+    }
+
+    private fun play(trackModels: ArrayList<Music?>?) {
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt]: chera bayad nullable bashe?
ya music darim play konim ya play nemikonim dige

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt [hunk 10]
```diff
@@ -0,0 +1,425 @@
...
+        reInit()
+        if (trackModels != null) {
+            val mediaSources =
+                arrayOfNulls<MediaSource>(trackModels.size)
+            for (i in trackModels.indices) {
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt:51]: ba `.forEach()` fek konam kheili behtar bashim

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt [hunk 11]
```diff
@@ -0,0 +1,425 @@
...
+
+    private fun pausePlayer() {
+        playbackPosition = player.currentPosition
+        currentWindow = player.currentWindowIndex
+        player.playWhenReady = false
+        showCurrentTrackNotification(musicList?.get(currentWindow))
+    }
+
+    private fun startPlayer() {
+        player.seekTo(currentWindow, playbackPosition)
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt]: chandin bar dare `seekTo` seda zade mishe
benazaram yek bar kafi bashe
aslan mage ba play niaz hast?

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt [hunk 12]
```diff
@@ -0,0 +1,425 @@
...
+
+
+    lateinit var mediaSource: ConcatenatingMediaSource
+
+    lateinit var player: SimpleExoPlayer
+
+    private var currentWindow = 0
+    private var playbackPosition: Long = 0
+    private var musicList: ArrayList<Music?>? = null
+    private var currentMusicId: Int = 0
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt]: in ba window che farghi dare?

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt [hunk 13]
```diff
@@ -0,0 +1,425 @@
...
+                NOTIFICATION_ID.FOREGROUND_SERVICE,
+                notificationBuilder.build()
+            )
+        }
+        return music
+    }
+
+    private fun releasePlayer() {
+//        if (player != null) {
+        playbackPosition = player.currentPosition
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt]: vaghti dare destroy mishe che fayde i dare negah dari ina?

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt [hunk 14]
```diff
@@ -0,0 +1,425 @@
...
+        currentWindow = player.currentWindowIndex
+        player.stop()
+        player.release()
+//            player = null
+//        }
+    }
+
+    override fun onBind(intent: Intent): IBinder? {
+        return null
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt:290]: benazaram niaze bind konim service ro ke betonim UI neshon bedim
albate mitonim bezarim baraye task e baadi

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt [hunk 15]
```diff
@@ -0,0 +1,425 @@
...
+//        sendBroadcast_stateIdle()
+    }
+
+    override fun onPositionDiscontinuity(reason: Int) {
+
+    }
+
+    override fun onPlaybackParametersChanged(playbackParameters: PlaybackParameters) {}
+    override fun onSeekProcessed() {}
+    interface ACTION {
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt]: chera interface sakhti?
age companion object mikahy ke esm dashte bashe intori ham mishe:
`companion object Action`
albate benazaram aslan niazi be companion nist

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt [hunk 16]
```diff
@@ -0,0 +1,425 @@
...
+            const val PLAY_ACTION = "action.play"
+            const val PAUSE_ACTION = "action.pause"
+            const val NEXT_ACTION = "action.next"
+            const val START_TRACK_ACTION = "action.start.track"
+            const val START_TRACKS_ARRAY_ACTION = "action.start.tracksArray"
+            const val STOP_ACTION = "action.stop"
+        }
+    }
+
+    interface NOTIFICATION_ID {
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt]: hamchenin

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt [hunk 17]
```diff
@@ -0,0 +1,425 @@
...
+    }
+
+
+    private fun generatePlayIntent(context: Context?): Intent {
+        val playIntent = Intent(context, ExoPlayerService::class.java)
+        playIntent.action = ACTION.PLAY_ACTION
+        return playIntent
+    }
+
+    private fun generatePauseIntent(context: Context?): Intent {
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt]: benazaram behtare baraye ina ye `factory` neveshte beshe
albate ba koltin mishe khoshgel tar ham nevesht!

**Review by helloseyedjafari — CHANGES_REQUESTED** [review]
right track 
but improvements needed

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt (hunk 8 above)
Comment (mahdi302) [reply app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt:54]: albateh to injam faghat yebar sakhte mishe

File: app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListFragment.kt (hunk 6 above)
Comment (mahdi302) [reply app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListFragment.kt:147]: are testi gozashte shode

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt (hunk 17 above)
Comment (mahdi302) [reply app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt]: ok

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt (hunk 16 above)
Comment (mahdi302) [reply app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt]: ok

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt (hunk 15 above)
Comment (mahdi302) [reply app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt]: ok

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt (hunk 11 above)
Comment (mahdi302) [reply app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt:86]: in method 2 bar seda zade mishe :
1-user bare aval ro ye music click mikone(ke baraye tashkhise inke user kodum music ro click karde ke pakhsh beshe niyaz )
 2-user music ro play mikone:(baraye tashkhise pakhshe music as jayee ke pause shode niyaz)

Comment (mahdi302) [reply app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt]: method startPlayer() vase action haye(next,prev,play) estefade shode ke baraye har kodum az in action ha niyaz hast player.seekTo seda zade beshe

File: app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListFragment.kt (hunk 6 above)
Comment (mahdi302) [reply app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListFragment.kt]: Are vase test gozashtam

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt (hunk 8 above)
Comment (mahdi302) [reply app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt]: ok

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt (hunk 9 above)
Comment (mahdi302) [reply app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt]: dorostesh mikonam

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt (hunk 12 above)
Comment (mahdi302) [reply app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt]: dorostesh mikonam

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt (hunk 13 above)
Comment (mahdi302) [reply app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt]: dorostesh mikonam

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt (hunk 14 above)
Comment (mahdi302) [reply app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt:290]: ok

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt (hunk 15 above)
Comment (mahdi302) [reply app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt]: dorostesh mikonam

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt (hunk 16 above)
Comment (mahdi302) [reply app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt]: dorostesh mikonam

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt (hunk 17 above)
Comment (mahdi302) [reply app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt]: dorostesh mikonam

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt (hunk 10 above)
Comment (mahdi302) [reply app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt:51]: be i vase add kardan to mediaSources  niyaz hast

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt (hunk 7 above)
Comment (mahdi302) [reply app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt]: dorostesh mikonam

File: app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListFragment.kt (hunk 3 above)
Comment (mahdi302) [reply app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListFragment.kt]: movaghat gozashtam

File: app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListFragment.kt (hunk 4 above)
Comment (mahdi302) [reply app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListFragment.kt]: movagat ghzashtam

File: app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListFragment.kt (hunk 5 above)
Comment (mahdi302) [reply app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListFragment.kt]: zamani bayad list ro pass bedim be  service ke user ru ye music click kone

File: app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListFragment.kt (hunk 3 above)
Comment (helloseyedjafari) [reply app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListFragment.kt]: khob key eslah mishe?
PR merge she nemishe taghyiresh dad.
va benazaram ravande aslie app bayad ghabl az merge in PR kamle va dorost bashe

File: app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListFragment.kt (hunk 5 above)
Comment (helloseyedjafari) [reply app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListFragment.kt]: soal e man chize dige i bod.
man naporsidam key pas midim goftam chera pass midim

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt [hunk 18]
```diff
@@ -0,0 +1,376 @@
...
+        player = SimpleExoPlayer.Builder(this).build()
+        player.setAudioAttributes(
+            com.google.android.exoplayer2.audio.AudioAttributes.Builder()
+                .setUsage(C.USAGE_MEDIA)
+                .setContentType(C.CONTENT_TYPE_SPEECH)
+                .build(), true
+        )
+
+//        player.seekTo(currentWindow, playbackPosition)
+//        player.playWhenReady = true
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt]: age niazi nist lotfan comment haro haz kon

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt [hunk 19]
```diff
@@ -0,0 +1,376 @@
...
+                if (next >= 0 && next < mediaSource.size) {
+                    currentMusicPosition = next
+                } else {
+                    currentMusicPosition = defaultValue
+                    playbackPosition = defaultValue.toLong()
+                }
+                startPlayer()
+            }
+            STOP_ACTION -> stopAction()
+        }
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt:158]: benazaram ro halate `else` throw konim ke jeloye eshtebahat ayandaro begirim

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt [hunk 20]
```diff
@@ -0,0 +1,376 @@
...
+        private const val PREV_ACTION = "action.prev"
+        private const val PLAY_ACTION = "action.play"
+        private const val PAUSE_ACTION = "action.pause"
+        private const val NEXT_ACTION = "action.next"
+        private const val START_TRACK_ACTION = "action.start.track"
+        private const val START_TRACKS_ARRAY_ACTION = "action.start.tracksArray"
+        private const val STOP_ACTION = "action.stop"
+        private const val CURRENT_MUSIC_POSITION_KEY = "currentMusicPosition"
+        private const val FOREGROUND_SERVICE_NOTIFICATION_ID = 101
+        private const val defaultValue = 0
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt]: mafhomi nadare in ziad
bezar jeloye hamon maghadir benazaram behtrare

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt [hunk 21]
```diff
@@ -0,0 +1,376 @@
...
+
+
+        fun actionStart(
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt:362]: benazaram bad nist baraye tak music ham function dashte bashim

File: app/src/main/java/com/worldsnas/starplayer/model/MusicRepoModel.kt [hunk 22]
```diff
@@ -1,10 +1,14 @@
 package com.worldsnas.starplayer.model
 
+import android.os.Parcelable
+import kotlinx.android.parcel.Parcelize
+
+@Parcelize
 data class MusicRepoModel(
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/model/MusicRepoModel.kt:4]: fek konam class e `music` parcelable hast
benazaram `RepoModel` nabayad parcelable bashe

**Review by helloseyedjafari — CHANGES_REQUESTED** [review]
progress. but small changes to go

File: app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListFragment.kt (hunk 5 above)
Comment (mahdi302) [reply app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListFragment.kt]: khob age bekhaim kole ahanghaye list pakhsh she bayad kolesh ro pass bedim be service.
age rahe dige hast pishnahad bedin. merci

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt [hunk 23]
```diff
@@ -0,0 +1,387 @@
...
+
+    private var playbackPosition: Long = 0
+    private lateinit var musicList: ArrayList<Music>
+    private var currentMusicPosition: Int = 0
+
+
+    override fun onDestroy() {
+        super.onDestroy()
+        releasePlayer()
+        releasePlayer()
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt]: duplicate?

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt [hunk 24]
```diff
@@ -0,0 +1,387 @@
...
+                } else {
+                    currentMusicPosition = 0
+                    playbackPosition = 0
+                }
+                startPlayer()
+            }
+            STOP_ACTION -> {
+                stopAction()
+            }
+            else -> stopAction()
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt]: benazaram throw konim behtare
sari tar moshkelo peyda mikonim
(albate maghadir ro ham pas bedim be exception)

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt [hunk 25]
```diff
@@ -0,0 +1,387 @@
...
+                putExtra(CURRENT_MUSIC_POSITION_KEY, currentMusicPosition)
+            })
+
+        }
+
+        fun actionStart(
+            context: Context?,
+            music: Music
+        ) {
+            if (context == null) return
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt]: age nullable ghabol nemikonim behtare function nullable nabashe

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt [hunk 26]
```diff
@@ -0,0 +1,387 @@
...
+        private const val NEXT_ACTION = "action.next"
+        private const val START_TRACK_ACTION = "action.start.track"
+        private const val START_TRACKS_ARRAY_ACTION = "action.start.tracksArray"
+        private const val STOP_ACTION = "action.stop"
+        private const val CURRENT_MUSIC_POSITION_KEY = "currentMusicPosition"
+        private const val FOREGROUND_SERVICE_NOTIFICATION_ID = 101
+
+
+        fun actionStart(
+            context: Context?,
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt]: inja ham hamintor

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt [hunk 27]
```diff
@@ -0,0 +1,387 @@
...
+
+
+    private fun generateIntent(context: Context, action: String) =
+        Intent(context, ExoPlayerService::class.java).apply {
+            this.action = action
+        }
+
+
+    companion object {
+        private const val TAG = "PlayerService"
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt]: esme khobi nist benazaram

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt [hunk 28]
```diff
@@ -0,0 +1,387 @@
...
+
+    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
+
+        when (intent.action) {
+            PLAY_ACTION -> {
+
+                startPlayer()
+            }
+            PAUSE_ACTION -> pausePlayer()
+            START_TRACKS_ARRAY_ACTION -> {
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt:115]: in ba list esh hich farghi nadare benazaram to factory function ha handle konim behtar bashe nist?

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt [hunk 29]
```diff
@@ -0,0 +1,387 @@
...
+
+
+    private fun play(trackModel: Music) {
+        musicList = ArrayList<Music>()
+        musicList.add(trackModel)
+        play(musicList)
+    }
+
+    private fun play(trackModels: ArrayList<Music>) {
+        playbackPosition = 0
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt:46]: behtar nist ghabl az pakhsh yedor state ghabli ro chek konim?
on media source hayi ke az ghabl hast ro pak ya haz konim (nemidonam aslan niaz has ya na)

**Review by helloseyedjafari — CHANGES_REQUESTED** [review]
Good progress

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt (hunk 28 above)
Comment (mehdirouhandeh) [reply app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt:115]: manzureto motevageh nashodam

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt (hunk 29 above)
Comment (mehdirouhandeh) [reply app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt:46]: in method daghighan bad az ejra shodan service va initialize shodan player call mishe.
fekr nakonam niaz bashe vali mitooniam check konim ke age player dar hale pakhsh bood yebar release she va 2bare initialize she

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt (hunk 27 above)
Comment (mehdirouhandeh) [reply app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt]: esme ExoPlayerService khube ?

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt (hunk 25 above)
Comment (mehdirouhandeh) [reply app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt]: dorostesh mikonam

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt (hunk 24 above)
Comment (mehdirouhandeh) [reply app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt]: dorostesh mikonam

File: app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt (hunk 23 above)
Comment (mehdirouhandeh) [reply app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt]: dorostesh mikonam

**Review by helloseyedjafari — APPROVED** [review]

**Pull request merged**